README.md
# Facebook PHP SDK for CodeIgniter

Library for integration of Facebook PHP SDK v5+ with CodeIgniter 3+

**Version:** 3.0.0

> This library is meant to help you to use the Facebook PHP SDK in a simpler way with CodeIgniter. It is however not at a "can do everything" type of library. If you required highly advanced integrations with the Facebook PHP SDK I recommend you read up on the [documentation](https://developers.facebook.com/docs/reference/php) and use the Facebook PHP SDK directly.

## Requirements

- PHP 5.4+
- [CodeIgniter 3](http://www.codeigniter.com/)
- [CodeIgniter session library](http://www.codeigniter.com/userguide3/libraries/sessions.html)
- [Facebook PHP SDK v5](https://developers.facebook.com/docs/php/gettingstarted/5.0.0)
- [Composer](https://getcomposer.org/)

## Notice

Facebook *Canvas* and *Page Tab* support is experimental as I have not been able to test or confirm it working. If you test it, please report back if you had success or failure.

## Installation

1. Download the library files and add the files to your CodeIgniter installation. Only the library, config and composer.json files are required.
1. In your CodeIgniter `/application/config/config.php` file, set `$config['composer_autoload']` to `TRUE`. [Read more](https://www.codeigniter.com/user_guide/general/autoloader.html)
2. In your CodeIgniter `/application/config/config.php` file, configure the `Session Variables` section. [Read more](https://www.codeigniter.com/user_guide/libraries/sessions.html)
3. Update the `facebook.php` config file in `/application/config/facebook.php` with your Facebook App details.
4. Install the Facebook PHP SDK by navigating to your applications folder in the terminal and run Composer with `composer install`. [Read more](https://developers.facebook.com/docs/php/gettingstarted#install-composer)
6. Autoload the library in `application/config/autoload.php` or load it in needed controllers with `$this->load->library('facebook');`.
5. Enjoy!

## Usage

The library download includes a sample controller and views. The example code might not be the best or most beautiful code, but it is there to help you get started quicker.

## Methods

#### is_authenticated()

Check if user is authenticated. Returns access token object if user is, else empty.

```php
$this->facebook->is_authenticated();
```

#### login_url()

Get login url. This method will only return a URL when using the redirect (web) login method.

```php
$this->facebook->login_url();
```

#### logout_url()

Check if user is logged. This method will only return a URL when using the redirect (web) login method.

```php
$this->facebook->logout_url();
```

#### destroy_session()

Should only be used on the logout redirect url location. This method will unset the Facebook token cookie set by this library only. **This method can not be used to log out a user!**

```php
$this->facebook->destroy_session();
```

#### request(string $method, string $endpoint [, array $params [, string $access_token]])

Main method to do Graph requests. Should support all, or at least most of the avaialble Facebook Graph methods.

> *All methods has not been tested, please submit an issue report if you encounter any issues*

This method returns an array on both success or failure.

| Parameter | Type | Description |
| --- | --- | --- |
| $method | string | Request type. [get, post, delete] |
| $endpoint | string | Graph endpoint, eg /me to get user information. |
| $params | array | Array with extra graph parameters. This is optional. |
| $access_token | string | Optional access token. |

##### Example

Short exampel to get user information and print it out.

```php
$user = $this->facebook->request('get', '/me');
if (!isset($user['error']))
{
    print_r($user);
}
```

##### Notes

- Facebook Graph API documentation can be found [here](https://developers.facebook.com/docs/graph-api).
- List of Facebook Graph API Error codes can be found [here](https://developers.facebook.com/docs/graph-api/using-graph-api/v2.4#errors).

#### user_upload_request(string $path_to_file [, array $params [, string $type [, string $access_token]]])
Upload a image or video file to a users profile.

> This method only support upload to a users profile. Upload to a page is not possible at this moment.

| Parameter | Type | Description |
| --- | --- | --- |
| $path_to_file | string | Path to file on server or external URL |
| $params | array | Array with extra graph parameters. This is optional. |
| $type | string | Type of upload. [image, video] This is optional, Default: image |
| $access_token | string | Optional access token. |

##### Example

Short exampel to upload a file from the server.

```php
$this->facebook->user_upload_request('/path/to/file.jpg', ['message' => 'This is a test upload']);
```

##### Notes

- Facebook Graph API documentation can be found [here](https://developers.facebook.com/docs/graph-api).
- List of Facebook Graph API Error codes can be found [here](https://developers.facebook.com/docs/graph-api/using-graph-api/v2.4#errors).

#### add_to_batch_pool(string $key, string $method, string $endpoint [, array $params [, string $access_token = null]])

Add a request to the batch pool to be sent later in batch request.

| Parameter | Type | Description |
| --- | --- | --- |
| $key | string | Key for identifying this request in the batch response |
| $method | string | Request type. [get, post, delete] |
| $endpoint | string | Graph endpoint, eg /me to get user information. |
| $params | array | Array with extra graph parameters. This is optional. |
| $access_token | string | Optional access token. |

##### Example

```php
$this->facebook->add_to_batch_pool('user-profile', 'get', '/me');
```

#### remove_from_batch_pool(string $key)

Add a request to the batch pool to be sent later in batch request.

| Parameter | Type | Description |
| --- | --- | --- |
| $key | string | Key for identifying request to remove |

##### Example

```php
$this->facebook->remove_from_batch_pool('user-profile', 'get', '/me');
```

#### send_batch_pool()

Send the batch pool requests.

Example on how batch requests works in the [Facebook documentation](https://developers.facebook.com/docs/php/howto/example_batch_request).

##### Example

```php
$this->facebook->send_batch_pool();
```

##### Full example

```php
$this->facebook->add_to_batch_pool('user-profile', 'get', '/me');
$this->facebook->remove_from_batch_pool('user-email', 'get', '/me?fields=email');
$responses = $this->facebook->send_batch_pool();

foreach ($responses as $key => $data)
{
	print_r($key);
	print_r($data);
}
```

#### object()
If you want to work directly with the Facebook\Facebook service class, you can do so. The `object()` method will return the full object of `new Facebook\Facebook` service class that you can use however you would like.

> *The library will still take care of the loading of the SDK and load configured login helper.*

Documentation for Facebook\Facebook service class can be found [here](https://developers.facebook.com/docs/php/Facebook/5.0.0) and full SDK reference list [here](https://developers.facebook.com/docs/php/api/5.0.0).

##### Example

```php
$fb = $this->facebook->object();

// Get user info
$response = $fb->get('/me');
$user     = $response->getDecodedBody();

print_r($user);
```


changelog.md
# Changelog for Facebook PHP SDK for CodeIgniter

#### 3.0.0 (May 2016)
- NEW: New config setting `facebook_auth_on_load` to control on page load authorization (try to get valid access token). On as default for backward compatibility.
- NEW: New way of checking if access token in session is still valid. Should be more reliable now.
- NEW: New `user_upload_request()` method. Supports upload of images and video to a users profile.
- NEW: New `add_to_batch_pool()` method. Add multiple Graph request to the pool and send all in one go later.
- NEW: New `remove_from_batch_pool()` method. Remove a graph request from the batch pool.
- NEW: New `send_batch_pool()` method. Send all Graph request in the pool.
- UPDATE: Default graph version in config now v2.6
- UPDATE: Cleanup of the code
- CHANGE: Page load authorization (try to get valid access token) made optional but on as default. Control with new config settings.

#### 3.0.0-b02 (March 2016)
- FIX: Solved authentication problems by using `authenticate` method in `is_authenticated` method instead of checking only local session.
- UPDATE: Tested with Facebook PHP SDK v5.1.2
- UPDATE: Updated example code to use latest Graph API v2.5
- UPDATE: Updated example web login code to include users email.

#### 3.0.0-b01 (October 2015)
- NEW: Support for Facebook PHP SDK v5.
- NEW: Now using one singly dynamic method to do any type of Facebook Graph request.
- CHANGE: Dropped support for Facebook PHP SDK v4.

#### 2.0.0 (April 2015)
- NEW: Added support for Redirect login.
- NEW: Added support for Canvas login.
- CHANGE: Changed format of most methods return data. Now include status codes.
- UPDATE: Updated example code.

#### 1.1.0 (April 2015)
- First public release of library


application/views/example.php
<!DOCTYPE html>
<html>
<head>
<title>Facebook Login JavaScript Example</title>
<meta charset="UTF-8">
</head>
<body>
<script>
  // This is called with the results from from FB.getLoginStatus().
  function statusChangeCallback(response) {
    console.log('statusChangeCallback');
    console.log(response);
    // The response object is returned with a status field that lets the
    // app know the current login status of the person.
    // Full docs on the response object can be found in the documentation
    // for FB.getLoginStatus().
    if (response.status === 'connected') {
      testAPI();
    } else if (response.status === 'not_authorized') {
      // The person is logged into Facebook, but not your app.
      document.getElementById('status').innerHTML = 'Please log ' +
        'into this app.';
    } else {
      // The person is not logged into Facebook, so we're not sure if
      // they are logged into this app or not.
      document.getElementById('status').innerHTML = 'Please log ' +
        'into Facebook.';
    }
  }

  // This function is called when someone finishes with the Login
  // Button.  See the onlogin handler attached to it in the sample
  // code below.
  function checkLoginState() {
    FB.getLoginStatus(function(response) {
      statusChangeCallback(response);
    });
  }

  window.fbAsyncInit = function() {
  FB.init({
    appId      : '<?php echo $this->config->item('facebook_app_id'); ?>',
    cookie     : true,  // enable cookies to allow the server to access
                        // the session
    xfbml      : true,  // parse social plugins on this page
    version    : 'v2.2' // use version 2.2
  });

  // Now that we've initialized the JavaScript SDK, we call
  // FB.getLoginStatus().  This function gets the state of the
  // person visiting this page and can return one of three states to
  // the callback you provide.  They can be:
  //
  // 1. Logged into your app ('connected')
  // 2. Logged into Facebook, but not your app ('not_authorized')
  // 3. Not logged into Facebook and can't tell if they are logged into
  //    your app or not.
  //
  // These three cases are handled in the callback function.

  FB.getLoginStatus(function(response) {
    statusChangeCallback(response);
  });

  };

  // Load the SDK asynchronously
  (function(d, s, id) {
    var js, fjs = d.getElementsByTagName(s)[0];
    if (d.getElementById(id)) return;
    js = d.createElement(s); js.id = id;
    js.src = "//connect.facebook.net/en_US/sdk.js";
    fjs.parentNode.insertBefore(js, fjs);
  }(document, 'script', 'facebook-jssdk'));

  // Here we run a very simple test of the Graph API after login is
  // successful.  See statusChangeCallback() for when this call is made.
  function testAPI() {
    console.log('Welcome!  Fetching your information.... ');
    FB.api('/me', function(response) {
      console.log('Successful login for: ' + response.name);
      document.getElementById('status').innerHTML =
        'Thanks for logging in, ' + response.name + '! Refresh the page to update your details below.';
    });
  }
</script>

<!--
  Below we include the Login Button social plugin. This button uses
  the JavaScript SDK to present a graphical Login button that triggers
  the FB.login() function when clicked.
-->

<fb:login-button scope="public_profile,email" onlogin="checkLoginState();">
</fb:login-button>

<div id="status">
</div>

<p><a href="<?php echo $this->facebook->login_url(); ?>">Redirect login</a></p>
<p><a href="<?php echo $this->facebook->logout_url(); ?>">Redirect logout</a></p>

<p><?php echo 'Facebook User ID:' . $user_id; ?></p>
<p>
  <?php echo 'Facebook User details:';

  echo '<ul>';
  foreach ($user as $detail => $value)
  {
    if ($detail == 'permissions') {
      echo '<li>' . $detail .' : ';
        print_r($value);
      echo '</li>';
    }
    else
    {
      echo '<li>' . $detail .' : ' . $value . '</li>';
    }
  }
  echo '</ul>';

  ?>
</p>

</body>
</html>


application/config/facebook.php
<?php
defined('BASEPATH') OR exit('No direct script access allowed');

/*
| -------------------------------------------------------------------
|  Facebook App details
| -------------------------------------------------------------------
|
| To get an facebook app details you have to be a registered developer
| at http://developer.facebook.com and create an app for your project.
|
|  facebook_app_id               string   Your facebook app ID.
|  facebook_app_secret           string   Your facebook app secret.
|  facebook_login_type           string   Set login type. (web, js, canvas)
|  facebook_login_redirect_url   string   URL tor redirect back to after login. Do not include domain.
|  facebook_logout_redirect_url  string   URL tor redirect back to after login. Do not include domain.
|  facebook_permissions          array    The permissions you need.
|  facebook_graph_version        string   Set Facebook Graph version to be used. Eg v2.6
|  facebook_auth_on_load         boolean  Set to TRUE to have the library to check for valid access token on every page load.
*/

$config['facebook_app_id']              = '680409738751130';
$config['facebook_app_secret']          = '2c1cc39c1994aac6658ff57c64ea38a6';
$config['facebook_login_type']          = 'js';
$config['facebook_login_redirect_url']  = 'example/web_login';
$config['facebook_logout_redirect_url'] = 'example/logout';
$config['facebook_permissions']         = array('public_profile', 'publish_actions', 'email');
$config['facebook_graph_version']       = 'v2.6';
$config['facebook_auth_on_load']        = TRUE;


application/libraries/Facebook.php
<?php defined('BASEPATH') OR exit('No direct script access allowed');

/**
 * Facebook PHP SDK for CodeIgniter 3
 *
 * Library wrapper for Facebook PHP SDK. Check user login status, publish to feed
 * and more with easy to use CodeIgniter syntax.
 *
 * This library requires the Facebook PHP SDK to be installed with Composer, and that CodeIgniter
 * config is set to autoload the vendor folder. More information in the CodeIgniter user guide at
 * http://www.codeigniter.com/userguide3/general/autoloader.html?highlight=composer
 *
 * It also requires CodeIgniter session library to be correctly configured.
 *
 * @package     CodeIgniter
 * @category    Libraries
 * @author      Mattias Hedman
 * @license     MIT
 * @link        https://github.com/darkwhispering/facebook-sdk-v4-codeigniter
 * @version     3.0.0
 */

use Facebook\Facebook as FB;
use Facebook\Authentication\AccessToken;
use Facebook\Exceptions\FacebookResponseException;
use Facebook\Exceptions\FacebookSDKException;
use Facebook\FacebookBatchResponse;
use Facebook\Helpers\FacebookCanvasHelper;
use Facebook\Helpers\FacebookJavaScriptHelper;
use Facebook\Helpers\FacebookPageTabHelper;
use Facebook\Helpers\FacebookRedirectLoginHelper;

Class Facebook
{
    const UPLOAD_TYPE_VIDEO = 'video';
    const UPLOAD_TYPE_IMAGE = 'image';

    /**
     * @var FB
     */
    private $fb;

    /**
     * @var FacebookRedirectLoginHelper|FacebookCanvasHelper|FacebookJavaScriptHelper|FacebookPageTabHelper
     */
    private $helper;

    /**
     * @var array
     */
    private $batch_request_pool = [];

    /**
     * Facebook constructor.
     */
    public function __construct()
    {
        // Load config
        $this->load->config('facebook');

        // Load required libraries and helpers
        $this->load->library('session');
        $this->load->helper('url');

        if (!isset($this->fb))
        {
            $this->fb = new FB([
                'app_id'                => $this->config->item('facebook_app_id'),
                'app_secret'            => $this->config->item('facebook_app_secret'),
                'default_graph_version' => $this->config->item('facebook_graph_version')
            ]);
        }

        // Load correct helper depending on login type
        // set in the config file
        switch ($this->config->item('facebook_login_type'))
        {
            case 'js':
                $this->helper = $this->fb->getJavaScriptHelper();
                break;

            case 'canvas':
                $this->helper = $this->fb->getCanvasHelper();
                break;

            case 'page_tab':
                $this->helper = $this->fb->getPageTabHelper();
                break;

            case 'web':
                $this->helper = $this->fb->getRedirectLoginHelper();
                break;
        }

        // Try and authenticate the user right away
        $this->authenticate();
    }

    /**
     * @return FB
     */
    public function object()
    {
        return $this->fb;
    }

    /**
     * Check if user are logged in by checking if we have a Facebook
     * session active.
     *
     * @return mixed|boolean
     */
    public function is_authenticated()
    {
        $access_token = $this->authenticate();

        if (isset($access_token))
        {
            return $access_token;
        }

        return false;
    }

    /**
     * Do Graph request
     *
     * @param       $method
     * @param       $endpoint
     * @param array $params
     * @param null  $access_token
     *
     * @return array
     */
    public function request($method, $endpoint, $params = [], $access_token = null)
    {
        try
        {
            $response = $this->fb->{strtolower($method)}($endpoint, $params, $access_token);
            return $response->getDecodedBody();
        }
        catch(FacebookResponseException $e)
        {
            return $this->logError($e->getCode(), $e->getMessage());
        }
        catch (FacebookSDKException $e)
        {
            return $this->logError($e->getCode(), $e->getMessage());
        }
    }

    /**
     * Upload image or video to user profile
     *
     * @param        $path_to_file
     * @param array  $params
     * @param string $type
     * @param null   $access_token
     *
     * @return array
     */
    public function user_upload_request($path_to_file, $params = [], $type = self::UPLOAD_TYPE_IMAGE, $access_token = null)
    {
        if ($type === self::UPLOAD_TYPE_IMAGE)
        {
            $data = ['source' => $this->fb->fileToUpload($path_to_file)] + $params;
            $endpoint = '/me/photos';
        }
        elseif ($type === self::UPLOAD_TYPE_VIDEO)
        {
            $data = ['source' => $this->fb->videoToUpload($path_to_file)] + $params;
            $endpoint = '/me/videos';
        }
        else
        {
            return $this->logError(400, 'Invalid upload type');
        }

        try
        {
            $response = $this->fb->post($endpoint, $data, $access_token);
            return $response->getDecodedBody();
        }
        catch(FacebookSDKException $e)
        {
            return $this->logError($e->getCode(), $e->getMessage());
        }
    }

    /**
     * Add request to batch
     *
     * @param       $key
     * @param       $method
     * @param       $endpoint
     * @param array $params
     * @param null  $access_token
     */
    public function add_to_batch_pool($key, $method, $endpoint, $params = [], $access_token = null)
    {
        $this->batch_request_pool = array_merge(
            $this->batch_request_pool,
            [$key => $this->fb->request($method, $endpoint, $params, $access_token)]
        );
    }

    /**
     * Remove request from batch
     *
     * @param $key
     */
    public function remove_from_batch_pool($key)
    {
        if (isset($this->batch_request_pool[$key]))
        {
            unset($this->batch_request_pool[$key]);
        }
    }

    /**
     * Send all request in the batch pool
     *
     * @return array|FacebookBatchResponse
     */
    public function send_batch_pool()
    {
        try
        {
            $responses = $this->fb->sendBatchRequest($this->batch_request_pool);
            $this->batch_request_pool = [];

            $data = [];
            foreach ($responses as $key => $response)
            {
                $data[$key] = $response->getDecodedBody();
            }

            return $data;
        }
        catch(FacebookResponseException $e)
        {
            return $this->logError($e->getCode(), $e->getMessage());
        }
        catch(FacebookSDKException $e)
        {
            return $this->logError($e->getCode(), $e->getMessage());
        }
    }

    /**
     * Generate Facebook login url for Facebook Redirect Login (web)
     *
     * @return  string
     */
    public function login_url()
    {
        // Login type must be web, else return empty string
        if ($this->config->item('facebook_login_type') != 'web')
        {
            return '';
        }

        return $this->helper->getLoginUrl(
            base_url() . $this->config->item('facebook_login_redirect_url'),
            $this->config->item('facebook_permissions')
        );
    }

    /**
     * Generate Facebook login url for Facebook Redirect Login (web)
     *
     * @return string
     * @throws FacebookSDKException
     */
    public function logout_url()
    {
        // Login type must be web, else return empty string
        if ($this->config->item('facebook_login_type') != 'web')
        {
            return '';
        }

        // Create logout url
        return $this->helper->getLogoutUrl(
            $this->get_access_token(),
            base_url() . $this->config->item('facebook_logout_redirect_url')
        );
    }

    /**
     * Destroy our local Facebook session
     */
    public function destroy_session()
    {
        $this->session->unset_userdata('fb_access_token');
    }

    /**
     * Get a new access token from Facebook
     *
     * @return array|AccessToken|null|object|void
     */
    private function authenticate()
    {
        $access_token = $this->get_access_token();

        if ($access_token && $this->get_expire_time() > (time() + 30) || $access_token && !$this->get_expire_time())
        {
            $this->fb->setDefaultAccessToken($access_token);
            return $access_token;
        }

        // If we did not have a stored access token or if it has expired, try get a new access token
        if (!$access_token)
        {
            try
            {
                $access_token = $this->helper->getAccessToken();
            }
            catch (FacebookSDKException $e)
            {
                $this->logError($e->getCode(), $e->getMessage());
                return null;
            }

            // If we got a session we need to exchange it for a long lived session.
            if (isset($access_token))
            {
                $access_token = $this->long_lived_token($access_token);

                $this->set_expire_time($access_token->getExpiresAt());
                $this->set_access_token($access_token);
                $this->fb->setDefaultAccessToken($access_token);

                return $access_token;
            }
        }

        // Collect errors if any when using web redirect based login
        if ($this->config->item('facebook_login_type') === 'web')
        {
            if ($this->helper->getError())
            {
                // Collect error data
                $error = array(
                    'error'             => $this->helper->getError(),
                    'error_code'        => $this->helper->getErrorCode(),
                    'error_reason'      => $this->helper->getErrorReason(),
                    'error_description' => $this->helper->getErrorDescription()
                );

                return $error;
            }
        }

        return $access_token;
    }

    /**
     * Exchange short lived token for a long lived token
     *
     * @param AccessToken $access_token
     *
     * @return AccessToken|null
     */
    private function long_lived_token(AccessToken $access_token)
    {
        if (!$access_token->isLongLived())
        {
            $oauth2_client = $this->fb->getOAuth2Client();

            try
            {
                return $oauth2_client->getLongLivedAccessToken($access_token);
            }
            catch (FacebookSDKException $e)
            {
                $this->logError($e->getCode(), $e->getMessage());
                return null;
            }
        }

        return $access_token;
    }

    /**
     * Get stored access token
     *
     * @return mixed
     */
    private function get_access_token()
    {
        return $this->session->userdata('fb_access_token');
    }

    /**
     * Store access token
     *
     * @param AccessToken $access_token
     */
    private function set_access_token(AccessToken $access_token)
    {
        $this->session->set_userdata('fb_access_token', $access_token->getValue());
    }

    /**
     * @return mixed
     */
    private function get_expire_time()
    {
        return $this->session->userdata('fb_expire');
    }

    /**
     * @param DateTime $time
     */
    private function set_expire_time(DateTime $time = null)
    {
        if ($time) {
            $this->session->set_userdata('fb_expire', $time->getTimestamp());
        }
    }

    /**
     * @param $code
     * @param $message
     *
     * @return array
     */
    private function logError($code, $message)
    {
        log_message('error', '[FACEBOOK PHP SDK] code: ' . $code.' | message: '.$message);
        return ['error' => $code, 'message' => $message];
    }

    /**
     * Enables the use of CI super-global without having to define an extra variable.
     * I can't remember where I first saw this, so thank you if you are the original author.
     *
     * Borrowed from the Ion Auth library (http://benedmunds.com/ion_auth/)
     *
     * @param $var
     *
     * @return mixed
     */
    public function __get($var)
    {
        return get_instance()->$var;
    }


}


application/views/examples/web.php
<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <title>Facebook PHP SDK v4 for CodeIgniter - Redirect login example</title>

    <style>
        body {
            padding: 0;
            margin: 0;
            font-family: Helvetica, Sans-serif;
            font-size: 16px;
            color: #333;
            line-height: 1.5;
        }

        .wrapper {
            width: 800px;
            margin: 60px auto;
            border: 1px solid #eee;
            background: #fcfcfc;
            padding: 0 20px 20px;
            box-shadow: 1px 1px 2px rgba(0,0,0,0.1);
        }

        h1, h3 {
            text-align: center;
        }

        .login {
            text-align: center;
        }

        a {
            border: none;
            background: #2F5B85;
            color: #fff;
            font-size: 18px;
            padding: 10px 20px;
            margin: 20px auto;
            cursor: pointer;

            transition: background .6s ease;
        }

        a:hover {
            background: #999;
        }
    </style>
</head>
<body>

<div class="wrapper">

    <h1>Facebook PHP SDK v4 for CodeIgniter</h1>
    <h3>Redirect login example</h3>

    <p>
        Simple example how you can use the Facebook PHP SDK v4 for CodeIgniter and the Web Redirect login method.
    </p>

    <p>
        <strong>
            For this example to work, make sure you have set 'facebook_login_type' as 'web' and specified login and logout redirect links in the config file.
        </strong>
    </p>

    <p>
        This example code do 3 things
        <ol>
            <li>Check if a user is logged in on page load.</li>
            <li>If user are logged in, displayes some basic information about the user and a logout button.</li>
            <li>If user is not logged in, display login button.</li>
        </ol>
    </p>

    <?php if ( ! $this->facebook->is_authenticated()) : ?>

        <div class="login">
            <a href="<?php echo $this->facebook->login_url(); ?>">Login</a>
        </div>

    <?php else : ?>

        <div class="user-info">

            <p><strong>User information</strong></p>

            <ul>
                <?php foreach ($user as $key => $value) : ?>

                    <li><?php echo $key; ?> : <?php echo $value; ?></li>

                <?php endforeach; ?>
            </ul>

            <p>
                <a href="<?php echo $this->facebook->logout_url(); ?>">Logout</a>
            </p>

        </div>

		<div class="upload-form">
			<?php if ($this->input->post('imageFile')) {
				$response = $this->facebook->user_upload_request(
					$this->input->post('imageFile'),
					['message' => 'This is a test']
				);

				var_dump($response);
			} ?>

			<form method="POST">
				<input type="text" name="imageFile" value="<?php echo APPPATH; ?>MM8354_GrandCanyon.jpg"/>
				<input type="submit" name="upload" value="Upload"/>
			</form>
		</div>

    <?php endif; ?>

</div>

</body>
</html>


application/views/examples/js.php
<!DOCTYPE html>
<html>
<head>
	<meta charset="UTF-8">
	<title>Facebook PHP SDK for CodeIgniter - Javascript SDK login example</title>

	<style>
		body {
			padding: 0;
			margin: 0;
			font-family: Helvetica, Sans-serif;
			font-size: 16px;
			color: #333;
			line-height: 1.5;
		}

		.wrapper {
			width: 800px;
			margin: 60px auto;
			border: 1px solid #eee;
			background: #fcfcfc;
			padding: 0 20px 20px;
			box-shadow: 1px 1px 2px rgba(0,0,0,0.1);
		}

		h1, h3 {
			text-align: center;
		}

		.login,
		.form {
			text-align: center;
		}

		.login button,
		input[type="submit"] {
			border: none;
			background: #2F5B85;
			color: #fff;
			font-size: 18px;
			padding: 10px 20px;
			margin: 20px auto;
			cursor: pointer;

			transition: background .6s ease;
		}

		.login button:hover,
		input[type="submit"]:hover {
			background: #999;
		}

		textarea {
			width: 96%;
			height: 200px;
			background: #fff;
			border: 1px solid #ccc;
			padding: 2%;
		}

		.form {display:none;}

		.note {
			font-size: 12px;
			color: #888;
		}
	</style>
</head>
<body>

<div class="wrapper">

	<h1>Facebook PHP SDK for CodeIgniter</h1>
	<h3>Javascript SDK login example</h3>

	<p>Simple example how you can use the Facebook PHP SDK for CodeIgniter together with the Facebook Javascript SDK and the login to Facebook functionality.</p>

	<p><strong>For this example to work, make sure you have set 'facebook_login_type' as 'js' in the config file and have <i>publish_actions</i> permissions!</strong></p>

	<p>
		This example code do 4 things
		<ol>
			<li>Check if the user is logged in to Facebook on page load.</li>
			<li>If user are logged in, display form to user to publish to their wall.</li>
			<li>If user is not logged in, display login button.</li>
			<li>Display the form after login and publish to users wall when subbmitting form without any page refresh</li>
		</ol>
	</p>

	<div class="login">
		<button>Login</button>
	</div>

	<div class="form">
		<form class="post-to-wall">
			<textarea name="message" placeholder="Type some text here and submit to post to your wall"></textarea>
			<input type="submit" name="submit" value="Post" />
		</form>
	</div>

	<p class="note"><i>Note: You can publish text posts to a users wall using only the Javascript SDK. This is ONLY an example on how the Javascript SDK can work togheter with the PHP SDK to publish and/or read information and content.</i></p>


</div>

<script src="http://code.jquery.com/jquery-1.11.3.min.js"></script>
<script>
	// Initiate Facebook JS SDK
	window.fbAsyncInit = function() {
		FB.init({
			appId   : '<?php echo $this->config->item('facebook_app_id'); ?>', // Your app id
			cookie  : true,  // enable cookies to allow the server to access the session
			xfbml   : false,  // disable xfbml improves the page load time
			version : 'v2.4', // use version 2.4
			status  : true // Check for user login status right away
		});

		FB.getLoginStatus(function(response) {
			console.log('getLoginStatus', response);
			loginCheck(response);
		});
	};

	// Check login status
	function statusCheck(response)
	{
		console.log('statusCheck', response.status);
		if (response.status === 'connected')
		{
			$('.login').hide();
			$('.form').fadeIn();
		}
		else if (response.status === 'not_authorized')
		{
			// User logged into facebook, but not to our app.
		}
		else
		{
			// User not logged into Facebook.
		}
	}

	// Get login status
	function loginCheck()
	{
		FB.getLoginStatus(function(response) {
			console.log('loginCheck', response);
			statusCheck(response);
		});
	}

	// Here we run a very simple test of the Graph API after login is
	// successful.  See statusChangeCallback() for when this call is made.
	function getUser()
	{
		FB.api('/me', function(response) {
			console.log('getUser', response);
		});
	}

	$(function(){
		// Trigger login
		$('.login').on('click', 'button', function() {
			FB.login(function(){
				loginCheck();
			}, {scope: '<?php echo implode(",", $this->config->item('facebook_permissions')); ?>'});
		});

		$('.form').on('submit', '.post-to-wall', function(e) {
			e.preventDefault();

			var formdata = $(this).serialize();

			$.ajax({
				url: '/example/post',
				data: formdata,
				type: 'POST',
				dataType: 'json',
				success: function(response) {
					console.log(response);
					if (response.id)
					{
						$('.form').html('<p>Post submitted successfully.</p>');
					}
					else
					{
						$('.form').html('<p>Something happened, please try again!.</p>');
					}
				}

			})
		});
	});

	(function(d, s, id){
		var js, fjs = d.getElementsByTagName(s)[0];
		if (d.getElementById(id)) {return;}
		js = d.createElement(s); js.id = id;
		js.src = "//connect.facebook.net/en_US/sdk.js";
		fjs.parentNode.insertBefore(js, fjs);
	}(document, 'script', 'facebook-jssdk'));
</script>

</body>
</html>
